### PR TITLE
Add Fallback Option for Eth1 Nodes

### LINKED ISSUE #7969: Multiple ETH1 endpoints
# 🚀 Feature Request

### Description

Currently the beacon node accepts a single URL for ETH1 endpoint. Due to the recent AWS outage, its important to have multiple endpoint support just incase if the primary ETH1 fails, beacon node will then fallback to the secondary URL. Then if the primary endpoint recovers then start to switch from secondary to primary

### Describe the solution you'd like

Modify the beacon node code to accept multiple endpoints, and if the first URL fails, fallback to the second one.. Maintain a polling call to all endpoints for health checks. If they are healthy then recover to the first URL.

### Describe alternatives you've considered

As of now this is handled outside the application level which can be on NGINX / proxies. But its best to move this to application code.

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Addition

**What does this PR do? Why is it needed?**

- [x] Adds the ability to have fallbacks for our eth1 endpoints.
- [x] Fix references across the repo.
- [x] Performs regular health checks to switch over to primary node in the event it comes back up.
- [x] Add unit test to test this feature.

**Which issues(s) does this PR fix?**

Fixes #7969 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/flags/base.go [hunk 1]
```diff
@@ -9,10 +9,9 @@ import (
 
 var (
 	// HTTPWeb3ProviderFlag provides an HTTP access endpoint to an ETH 1.0 RPC.
-	HTTPWeb3ProviderFlag = &cli.StringFlag{
+	HTTPWeb3ProviderFlag = &cli.StringSliceFlag{
 		Name:  "http-web3provider",
 		Usage: "A mainchain web3 provider string http endpoint",
```
Comment (shayzluf) [inline beacon-chain/flags/base.go:14]:
```suggestion
		Usage: "A mainchain web3 provider string http endpoint. This flag may be used multiple times.",
```

File: beacon-chain/powchain/service.go [hunk 2]
```diff
@@ -817,3 +830,37 @@ func (s *Service) determineEarliestVotingBlock(ctx context.Context, followBlock
...
+	}
+	nextIndex := currIndex + 1
+	if nextIndex >= totalEndpoints {
+		nextIndex = 0
+	}
+	s.currHttpEndpoint = s.httpEndpoints[nextIndex]
+	log.Infof("Falling back to alternative endpoint: %s", s.currHttpEndpoint)
+}
+
+func dedupEndpoints(endpoints []string) []string {
```
Comment (shayzluf) [inline beacon-chain/powchain/service.go:855]: maybe add a test for this function?

File: shared/cmd/helpers.go [hunk 3]
```diff
@@ -74,24 +75,27 @@ func EnterPassword(confirmPassword bool, pr PasswordReader) (string, error) {
...
-	case strings.HasPrefix(web3endpoint, "ws://"):
-	case strings.HasPrefix(web3endpoint, "wss://"):
-	default:
-		web3endpoint, err := fileutil.ExpandPath(ctx.String(flag.Name))
-		if err != nil {
-			return errors.Wrapf(err, "could not expand path for %s", web3endpoint)
-		}
-		if err := ctx.Set(flag.Name, web3endpoint); err != nil {
-			return errors.Wrapf(err, "could not set %s to %s", flag.Name, web3endpoint)
+	rawFlags := sliceutil.SplitCommaSeparated(ctx.StringSlice(flags.Name))
```
Comment (shayzluf) [inline shared/cmd/helpers.go:83]: i think bootstrap nodes example uses a multiple flag and doesn't use SplitCommaSeparated	`c.BootstrapNodes = cliCtx.StringSlice(cmd.BootstrapNode.Name)`

File: shared/cmd/helpers.go [hunk 4]
```diff
@@ -74,24 +75,27 @@ func EnterPassword(confirmPassword bool, pr PasswordReader) (string, error) {
...
+	for i, rawValue := range rawFlags {
+		switch {
+		case strings.HasPrefix(rawValue, "http://"):
+		case strings.HasPrefix(rawValue, "https://"):
+		case strings.HasPrefix(rawValue, "ws://"):
+		case strings.HasPrefix(rawValue, "wss://"):
+		default:
+			web3endpoint, err := fileutil.ExpandPath(rawValue)
```
Comment (shayzluf) [inline shared/cmd/helpers.go:91]: not sure this code handles yaml file input
good example for this is in function `registerP2P` using `readbootNodes`

**Comment by mohamedmansour** [comment]
Thanks for doing this! The main issue with this design is that all the http endpoints are treated equally. We just want to provide a fallback endpoint in case the primary  endpoint stopped working. This design will only change current endpoint if a failure occurs.

Would be nice if there was a health check on the endpoints so that if the primary endpoint is back up, it will automatically switch back to that. So instead of doing that check if the call failed, maybe another worker should be maintaining the health status of every endpoint and setting its current url

**Comment by nisdas** [comment]
Hey @mohamedmansour , 

Thanks for giving your thoughts on this issue, performing a continous health check will require quite a bit of shifting around of our code. For obvious reasons the service as of this moment is only meant to be running a single persistent eth1 connection. I will think a bit more on if its feasible to be running continous health check on the primary eth1 node without messing around too much with our other core eth1 code.

As for the latter issue, it isnt relevant to the main purpose of the PR as of the current moment prysm only supports unauthenticated connections to eth1 nodes so it is out of scope for the PR. However support for JWT auth can be added in with a follow up PR.

**Comment by mohamedmansour** [comment]
@nisdas prysm currently supports cert based eth1 nodes. With the current design of the PR if we supply two different eth1 endpoints, we have to use the same cert which most likely isn’t the case.

Regarding my first question, it is very important for the fallback url not to become primary eth1 node for a long time if the primary eth1 node recovered. Otherwise this solution would only be valuable for those who only have strictly remote eth1 services like alchemy (which should be just a few).

The idea of introducing a failover is to be meerly be just temporal, but in this case it will be permanent (which is not what the issue on github is about).

Thanks for replying :)

**Comment by nisdas** [comment]
> prysm currently supports cert based eth1 nodes. With the current design of the PR if we supply two different eth1 endpoints, we have to use the same cert which most likely isn’t the case.

Currently prysm doesn't have the ability to provide a custom cert for an eth1 endpoint, we use the default certificate authorities to authenticate any upgraded https connections. From the view of the service there is no ability to provide a custom cert to an eth1 endpoint. This will also have to be designed along with JWT auth.

> Regarding my first question, it is very important for the fallback url not to become primary eth1 node for a long time if the primary eth1 node recovered. Otherwise this solution would only be valuable for those who only have strictly remote eth1 services like alchemy (which should be just a few).

Understood :) , will take a look at this a bit more deeply then :+1:

File: beacon-chain/powchain/service.go [hunk 5]
```diff
@@ -154,7 +158,7 @@ type Service struct {
 
 // Web3ServiceConfig defines a config struct for web3 service to use through its life cycle.
 type Web3ServiceConfig struct {
-	HTTPEndPoint       string
+	HTTPEndPoints      []string
```
Comment (rkapka) [inline beacon-chain/powchain/service.go:161]: Just for the sake of consistency: sometimes it's `endpoint/Endpoint` and sometimes it's `endPoint/EndPoint`. Can we unify this?

Comment (rkapka) [inline beacon-chain/powchain/service.go:161]: Just for the sake of consistency: sometimes it's `endpoint/Endpoint` and sometimes it's `endPoint/EndPoint`. Can we unify this?

**Review by rkapka — CHANGES_REQUESTED** [review]
I don't think changing the `HTTPWeb3ProviderFlag` to a slice is the best design. The problem is that if you specify the flag multiple times, it's not clear which endpoint is the primary one, and this is very important. Does the CLI library provide an ordering guarantee when using a slice flag?

A clearer approach IMO would be to keep the existing flag as it is, using it for the primary endpoint, and add a new `FallbackHTTPWeb3Provider(s)` flag for fallback endpoints, which can either be specified multiple times or as a comma-separated list of endpoints (adding `s` to the flag name in that case).

**Comment by rkapka** [comment]
Another option would be to allow specifying a `primary` suffix e.g. `http-web3provider=blah,primary` and default to `httpEndpoints[0]` in case no endpoint is marked as primary, maybe also issuing a warning in this case. We should also handle the scenario where multiple endpoints are marked as primary (error? choose the first one?)

File: shared/cmd/helpers.go [hunk 6]
```diff
@@ -73,8 +73,8 @@ func EnterPassword(confirmPassword bool, pr PasswordReader) (string, error) {
 	return passphrase, nil
 }
 
-// ExpandWeb3EndpointIfFile expands the path for --http-web3provider if specified as a file.
-func ExpandWeb3EndpointIfFile(ctx *cli.Context, flag *cli.StringFlag) error {
+// ExpandSingleEndpointIfFile expands the path for --http-web3provider if specified as a file.
```
Comment (rkapka) [inline shared/cmd/helpers.go:76]:
```suggestion
// ExpandSingleEndpointIfFile expands the path for --fallback-web3provider if specified as a file.
```

Comment (rkapka) [inline shared/cmd/helpers.go:76]:
```suggestion
// ExpandSingleEndpointIfFile expands the path for --fallback-web3provider if specified as a file.
```

File: shared/cmd/helpers.go [hunk 7]
```diff
@@ -96,3 +96,31 @@ func ExpandWeb3EndpointIfFile(ctx *cli.Context, flag *cli.StringFlag) error {
 	}
 	return nil
 }
+
+// ExpandWeb3EndpointsIfFile expands the path for --http-web3provider if specified as a file.
```
Comment (rkapka) [inline shared/cmd/helpers.go:100]:
```suggestion
// ExpandWeb3EndpointsIfFile expands the path for --fallback-web3provider if specified as a file.
```

File: shared/cmd/helpers_test.go [hunk 8]
```diff
@@ -80,33 +80,73 @@ func TestEnterPassword(t *testing.T) {
...
 	require.NoError(t, context.Set(HTTPWeb3ProviderFlag.Name, "./path.ipc"))
-	require.NoError(t, ExpandWeb3EndpointIfFile(context, HTTPWeb3ProviderFlag))
+	require.NoError(t, ExpandSingleEndpointIfFile(context, HTTPWeb3ProviderFlag))
 	require.Equal(t, curentdir+"/path.ipc", context.String(HTTPWeb3ProviderFlag.Name))
 }
+
+func TestExpandWeb3EndpointIfFile(t *testing.T) {
+	app := cli.App{}
+	set := flag.NewFlagSet("test", 0)
+	HTTPWeb3ProviderFlag := &cli.StringSliceFlag{Name: "http-web3provider", Value: cli.NewStringSlice()}
```
Comment (rkapka) [inline shared/cmd/helpers_test.go:117]:
```suggestion
	HTTPWeb3ProviderFlag := &cli.StringSliceFlag{Name: "fallback-web3provider", Value: cli.NewStringSlice()}
```

File: shared/cmd/helpers_test.go [hunk 9]
```diff
@@ -80,33 +80,73 @@ func TestEnterPassword(t *testing.T) {
...
 	// current dir path
 	curentdir, err := os.Getwd()
 	require.NoError(t, err)
 	require.NoError(t, context.Set(HTTPWeb3ProviderFlag.Name, "./path.ipc"))
-	require.NoError(t, ExpandWeb3EndpointIfFile(context, HTTPWeb3ProviderFlag))
+	require.NoError(t, ExpandSingleEndpointIfFile(context, HTTPWeb3ProviderFlag))
 	require.Equal(t, curentdir+"/path.ipc", context.String(HTTPWeb3ProviderFlag.Name))
 }
+
+func TestExpandWeb3EndpointIfFile(t *testing.T) {
```
Comment (rkapka) [inline shared/cmd/helpers_test.go:114]:
```suggestion
func TestExpandWeb3EndpointsIfFile(t *testing.T) {
```

**Comment by rkapka** [comment]
What about https://github.com/prysmaticlabs/prysm/pull/8062#discussion_r540090442? Do you think it's worthwhile to unify this?

File: beacon-chain/powchain/service.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/powchain/service.go]: Done

**Pull request merged**